pubtools/_pulp/arguments.py

```diff
@@ -43,3 +49,100 @@ def __call__(self, value):
+
+    To incorporate this action, set ``action=SplitAndExtend`` in
+    the call to ``ArgumentParser#add_argument``. By default this will
+    handle comma-delimited string args.
+
+    To use a different delimiter, pass ``split_on="your-delimiter"`` in
+    addition to ``action=SplitAndExtend``.
+
+    Examples:
+        >>> # setup the parser
+        >>> import sys
+        >>> from argparse import ArgumentParser
+        >>> parser = ArgumentParser()
+        >>> parser.add_argument("--option", type=str, action=SplitAndExtend, split_on=",")
+
+        Option with multiple instances, some are comma-delimited lists:
+        >>> sys.argv = ["command", "--option", "value1,value2", "--option", "value3"]
+        >>> args = parser.parse_args()
+        >>> print(args)
+        Namespace(option=['value1', 'value2', 'value3'])
+
+        Option with single instance, single value:
+        >>> sys.argv = ["command", "--option", "value1"]
+        >>> args = parser.parse_args()
+        >>> print(args)
+        Namespace(option=["value1"])
+
+        No option present:
+        >>> sys.argv = ["command"]
+        >>> args = parser.parse_args()
+        >>> print(args)
+        Namespace(option=None)
+
+        Option present, value missing
+        >>> sys.argv = ["command", "--option"]
+        >>> args = parser.parse_args()
+        usage: pydevconsole.py [-h] [--option OPTION]
+        pydevconsole.py: error: argument --option: expected one argument
+
+    Attributes:
+        split_on (str): the delimiter on which to split a delimited list
+            of vaules for a single instance of an option.
+
+    See Also:
+        `Built-in Argparse Actions`_
+            The set of built-in argparse Actions.
+
+    .. _Built-in Argparse Actions:
+        https://docs.python.org/3/library/argparse.html#action
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.__split_on = kwargs.pop("split_on", ",")
+        super(SplitAndExtend, self).__init__(*args, **kwargs)
+
+    def __call__(self, _, namespace, values, options=None):
+        items = getattr(namespace, self.dest, None) or []
+        # if values isn't a string, don't try to split it
+        # just add it to the accumulated list.
+
+        # by default, argparse will parse each value as a string,
+        # so unless this action is being used in conjunction with
+        # parser.add_argument(type=<some non-string type>) this
+        # should not be the case.
+        split = values.split(self.split_on) if isinstance(values, _STRING) else values
+
+        # it is possible to end up with "" as an entry
+        # after splitting in cases with an extra trailing
+        # delimiter like ``--option value1,`` - remove them.
+
+        # calling filter(None, Iterable) is equivalent to
+        # filter(lambda item: bool(item), Iterable)
+        # i.e. yields items which evaluate to True.
+        split = list(filter(None, split))
```

Comment:
Why do we do this filtering?

```
# (1)
"value1," => ["value1"]
"a,,b" => ["a", "b"]

# (2)
"value1," => ["value1", ""]
"a,,b" => ["a", "", "b"]
```

As I understand it, this implements (1).  Is there a reason the behavior in (1) is preferred over (2)?

I feel it's the other way around - (1) tries to guess what the caller probably meant, which is dangerous as the guess can be wrong.  (2) accurately maintains the data passed in by the caller, who may have had a valid reason for passing an empty string.  Generic code like this can't know that "silently ignore empty arguments" is the right thing to do for all use-cases.

If this was inspired by the `[x for x in type_strs if x] or None` in content_type which is being replaced - I think that code was just a way of making it return None if no argument was provided, it wasn't a goal to filter out empty strings passed by the caller.

Reply:
> perhaps removing the filtering is the correct thing to do overall

Yeah, can we please do that?  I'm concerned the filtering will have unintended consequences in the long term.

> removing the filter statement would introduce the possibility of empty strings in the list of content types going forward

I'm not concerned about that - I'm pretty certain that, in the previous code:

```python
        type_strs = (self.args.content_type or "").split(",")
        return [x for x in type_strs if x] or None
```

The filtering of empty values was just a shortcut to deal with the fact that the default value calculated by the first line would end up as `[""]` rather than `[]`.  If callers actually pass empty strings through here, I'd say it's acceptable to crash.

Reply:
Filtering empty strings has been removed.